### PR TITLE
Add range bucket iterator

### LINKED ISSUE #318: Add iterator() to RangeBucket
**Is your feature request related to a problem? Please describe.**
RangeBucket currently has not implemented `iterator()` which could be useful.

**Describe the solution you'd like**
Calling iterator we return an iterator which handles the conversion into RangeEntries automatically (it doesn't iterate over raw key/values as they must be processed by RangeDB to be more human readable).

### PR DESCRIPTION
**Pull request by karlfloersch**
## Description
This adds an `iterator()` function to the RangeBucket. This iterator is a bit custom because it has a function `nextRange()` which instead of returning a raw KV result, it instead parses it and returns a `RangeEntry`.

This feature is particularly useful if we ever want to iterate over the full contents of a RangeDB and return proper `RangeEntries`. I needed this for the way I was going to build the merkle tree creator using LevelDB. However, realized that our blocks won't get large enough to need to use a DB like this and so have abandoned that. Still this could be a useful feature!

## Questions
- Yall know of any cleaner ways to return this iterator other than extending our BaseIterator class?
- Should we make `next()` return a Range by default instead of calling it `nextRange()`? And maybe add `nextRaw()` which acts like the normal next function?

### Fixes
- Fixes #318

## Contributing Agreement

- [x] I have read and understood the [Plasma Group Contributing Guide and Code of Conduct](https://github.com/plasma-group/pigi/blob/master/.github/CONTRIBUTING.md) and am following those guidelines in this pull request.

### CODE REVIEW

File: packages/core/src/app/db/range-bucket.ts [hunk 1]
```diff
@@ -282,6 +284,23 @@ export class BaseRangeBucket implements RangeBucket {
...
+   * Creates an iterator with some options.
+   * @param options Parameters for the iterator.
+   * @returns the iterator instance.
+   */
+  public iterator(options?: IteratorOptions): RangeIterator {
+    const rangeIterator = new BaseRangeIterator(
+      this.db,
+      {
+        ...options,
+        prefix: this.addPrefix(this.prefix),
```
Comment (willmeister) [inline packages/core/src/app/db/range-bucket.ts]: Should this just be `this.prefix`, or do we want the prefix twice?

File: packages/core/test/app/db/range-bucket.spec.ts [hunk 2]
```diff
@@ -45,6 +45,22 @@ const testPutResults = async (
   db: RangeBucket,
   putContents: any[],
   expectedResults: any[]
+): Promise<void> => {
+  // First put the ranges
+  putRanges(db, putContents)
+  // Now check that they were added correctly
+  const res = await db.get(
+    new BigNum('0', 'hex'),
```
Comment (willmeister) [inline packages/core/test/app/db/range-bucket.spec.ts]: Can use `ZERO`

File: packages/core/test/app/db/range-bucket.spec.ts [hunk 3]
```diff
@@ -256,4 +269,32 @@ describe('RangeDB', () => {
       ]
     )
   })
+
+  describe('iterator()', () => {
+    it('allows nextRange() to be called by the iterator returning a RangeEntry instead of a KV', async () => {
+      const testRanges = {
```
Comment (willmeister) [inline packages/core/test/app/db/range-bucket.spec.ts]: I might suggest having the inputs be the expected results. If `db.put(...)` doesn't properly  overwrite range 50-150, this test will fail, even though the `iterator()` call works properly.

File: packages/core/src/app/db/range-bucket.ts (hunk 1 above)
Comment (karlfloersch) [reply packages/core/src/app/db/range-bucket.ts]: Ah good catch! This should be `options.prefix` not `this.prefix` as it is here -- https://github.com/plasma-group/pigi/blob/master/packages/core/src/app/db/bucket.ts#L80-L85

File: packages/core/test/app/db/range-bucket.spec.ts (hunk 2 above)
Comment (karlfloersch) [reply packages/core/test/app/db/range-bucket.spec.ts]: True, though there are a lot of `'0'`s which I'd have to change in this file. Do you think it's worth it to define a special `ZERO` constant?

**Pull request merged**